### PR TITLE
Split test_c10d.py to test_c10d_common.py, test_c10d_gloo.py, test_c10d_nccl.py

### PR DESCRIPTION
**Pull request by pbelevich**
Test Plan: NA

Differential Revision: D27913170

### CODE REVIEW

**Comment by SciPioneer** [comment]
Thanks for the refactoring! This will definitely make the test codebase manageable.

On the other hand, I think this PR does a bit more than plain splitting. Therefore, it will be better to add some descriptions in the PR to highlight some changes, e.g., creating some abstract classes like `AbstractCommTest`, `AbstractDistributedDataParallelTest` and `AbstractTimeoutTest`. 
 
 Another suggestion is separating out many fundamental store-related unit tests into a separate file, as these tests and other distributed training tests target different layers.

File: test/distributed/test_c10d_gloo.py [hunk 1]
```diff
@@ -0,0 +1,2130 @@
...
+            c10d.ReduceOp.MAX,
+            [torch.tensor([2 * rank + 1.0]), torch.tensor([2 * rank + 2.0])],
+            torch.tensor([2 * world_size]),
+        ),
+    ]
+
+
+class RendezvousEnvTest(TestCase):
+    @retry_on_connect_failures
+    def test_logging_init(self):
```
Comment (SciPioneer) [inline test/distributed/test_c10d_gloo.py:173]: Why is this test in gloo rather than common?

File: test/distributed/test_c10d_gloo.py [hunk 2]
```diff
@@ -0,0 +1,2130 @@
...
+
+        t1 = torch.zeros([1], dtype=torch.float32)
+        t2 = torch.zeros([1], dtype=torch.float64)
+        t3 = torch.zeros([2], dtype=torch.float32)
+
+        with self.assertRaisesRegex(ValueError, "requires non-empty input tensor list"):
+            pg.allgather([], [])
+
+        with self.assertRaisesRegex(
+                ValueError, "requires input/output tensor lists to have the same length"
```
Comment (SciPioneer) [inline test/distributed/test_c10d_gloo.py:907]: The indentation here and many "ValueError" lines are changed. Previously it's aligned with the end of "with". Is this intentional?

File: test/distributed/test_c10d_gloo.py [hunk 3]
```diff
@@ -0,0 +1,2130 @@
...
+        model_withload.load_state_dict(ddp_state_dict)
+
+        train_loop(ddp_withload, optimizer_withload, 3)
+        train_loop(model_withload, optimizer_non_ddp_withload, 3)
+
+        # re-run the model with the same inputs for 6 iterations with no checkpoint
+        train_loop(ddp_withoutload, optimizer_withoutload, 6)
+
+        for p_withload, p_withoutload, p_non_ddp_withload in zip(
+                ddp_withload.parameters(), ddp_withoutload.parameters(), model_withload.parameters()
```
Comment (SciPioneer) [inline test/distributed/test_c10d_gloo.py:1627]: Indentation changed.

**Review by wayi1 — APPROVED** [review]

File: test/distributed/test_c10d_gloo.py (hunk 1 above)
Comment (pbelevich) [reply test/distributed/test_c10d_gloo.py:173]: line 181:
```
        c10d.init_process_group(backend="gloo", init_method="env://")
```

Comment (SciPioneer) [reply test/distributed/test_c10d_gloo.py:172]: Better to add `@requires_gloo()` here.

**Pull request closed without merging**